### PR TITLE
Internal support of an alternative exception handler stack

### PR DESCRIPTION
**Pull request by mingweishih**
This PR introduces an internal mechanism that allows the OE runtime to set up a temporary, alternative stack for exception handlers. Such use cases include creating a pthread (see #4230 for example) with a small stack. The alternative stack allows the thread to handle a stack overflow exception, which is not supported by the default behavior where the exception handler directly uses the stack pointer saved on the SSA (in this case, pointing to invalid memory).  This PR also adds a new test that demonstrates how the stack overflow exception can be captured (required SGX2 hardware with in-enclave #PF/#GP delivery).

To use the API, OE runtime needs to be responsible to maintain the temporary stack by itself, and carefully set/unset the exception handler stack at proper time. The logic implemented by this PR always validates the stack pointer before using it. If the stack pointer is invalid, the logic will fallback to default behavior.

Signed-off-by: Ming-Wei Shih <mishih@microsoft.com>

### CODE REVIEW

**Comment by mingweishih** [comment]
bors try

**Comment by mingweishih** [comment]
bors try

**Comment by mingweishih** [comment]
bors try

**Comment by mingweishih** [comment]
> While this is a reasonable implementation, it would be good to consider alternatives.
> 
> * We could remember the last rsp, rbp values in OE stack before the stack is swapped by Mystikos. The exception handler can then use last oe rsp - ABI_RED_ZONE as its stack.
>   Mystikos reserves 2 pages of OE stack per thread. There is likely enough stack space left to run an exception handler.
> * We could reserve a portion of the top of the OE stack for running exception handlers.

I think this idea can also be enabled by this PR.
- In either case, the enter flow needs to determine which stack to use
   - Default: the rsp saved in SSA
   - This PR: `td->stack_handler_exception_handler_stack`
  
The Mystikos runtime can set `td->stack_handler_exception_handler_stack` to the OE stack frame - ABI_RED_ZONE before switching to Mystikos stack.
`oe_sgx_set_td_exception_handler_stack` can be used to set the `td->stack_handler_exception_handler_stack`.

File: enclave/core/sgx/enter.S [hunk 1]
```diff
@@ -105,15 +105,36 @@ oe_enter:
...
     // memory range so we do not need additional checks.
 
-.exception_check_pass:
+.exception_handler_stack_check:
+    // Stop speculative execution at target of conditional jump
+    lfence
+
+    // Get the exception_handler_stack_check range
+    mov td_exception_handler_stack(%r11), %r14
+    mov td_exception_handler_stack_size(%r11), %r15
```
Comment (anakrish) [inline enclave/core/sgx/enter.S:122]: If exception is triggered after exception stack is set, but before stack size is set, then r14 will be valid, but r15 will be zero. This will cause the exception handler to execute with RSP = bottom of exception stack, leading to memory corruption.

File: tests/VectorException/enc/init.cpp [hunk 2]
```diff
@@ -15,47 +18,62 @@ extern "C" void get_cpuid(
...
-    OE_TEST(c != 0);
-    OE_TEST(hits1 == 1);
-    OE_TEST(hits2 == 3);
+    int index = (use_exception_handler_stack) ? 1 : 0;
+    void* stack;
+
+    stack = malloc(EXCEPTION_HANDLER_STACK_SIZE);
+    if (!stack)
+        return;
+    oe_sgx_set_td_exception_handler_stack(stack, EXCEPTION_HANDLER_STACK_SIZE);
```
Comment (anakrish) [inline tests/VectorException/enc/init.cpp:67]: Not clear what this test is doing.  Between line 67 and 77, no exception is raised; so it doesn't matter whether exception stack is set or not.

File: enclave/core/sgx/td.c [hunk 3]
```diff
@@ -168,6 +168,23 @@ oe_sgx_td_t* oe_sgx_get_td()
...
+** oe_sgx_set_td_exception_handler_stack()
+**
+**     Internal API that allows an enclave to setup stack area for
+**     exception handlers to use.
+**
+**==============================================================================
+*/
+void oe_sgx_set_td_exception_handler_stack(void* stack, uint64_t size)
+{
+    oe_sgx_td_t* td = oe_sgx_get_td();
```
Comment (anakrish) [inline enclave/core/sgx/td.c:183]: It would be good to ensure that stack+size is 16 byte aligned even though oe_enter aligns the stack.
It makes the expectation clear.

File: enclave/core/sgx/enter.S [hunk 4]
```diff
@@ -105,15 +105,36 @@ oe_enter:
...
+    // Check if the stack range is within the enclave memory range
+    // If the check fails, fallback to the default behavior (i.e.,
+    // re-using the stack pointer saved in the SSA)
+    cmp %r12, %r14
+    jb .exception_stack_setup
+    cmp %r13, %r15
+    ja .exception_stack_setup
+
+    // Check passes, use the exception handler stack
+    mov %r15, %r8
```
Comment (anakrish) [inline enclave/core/sgx/enter.S:135]: ABI_REDZONE_BYTES at the top of the exception stack are never used.

File: enclave/core/sgx/enter.S (hunk 1 above)
Comment (mingweishih) [reply enclave/core/sgx/enter.S:122]: This seems not an issue as r14 and r15 will not be used to set rsp before the checks here are done.

Comment (anakrish) [reply enclave/core/sgx/enter.S:122]: Which are the checks you are referring to?
The checks in line 129 and 131 would both pass.

**Review by anakrish — COMMENTED** [review]
I'd suggest that we lock down the following as well in tests:
- RSP ought to be in the expected range when an exception handler is called. 
  Here are some scenarios I can think of, but there are others:
  - RSP must be in OE stack range if no exception stack has been set.
  - RSP must be in exception stack range if an exception stack has been set.
- Ocalls from within an exception handler ought to work correctly when exception stack has been installed.

File: enclave/core/sgx/enter.S (hunk 1 above)
Comment (mingweishih) [reply enclave/core/sgx/enter.S:122]: changed the order so the size is set first

File: tests/VectorException/enc/init.cpp (hunk 2 above)
Comment (mingweishih) [reply tests/VectorException/enc/init.cpp:62]: updated

**Comment by mingweishih** [comment]
> I'd suggest that we lock down the following as well in tests:
> 
> * RSP ought to be in the expected range when an exception handler is called.
>   Here are some scenarios I can think of, but there are others:
>   
>   * RSP must be in OE stack range if no exception stack has been set.
>   * RSP must be in exception stack range if an exception stack has been set.
> * Ocalls from within an exception handler ought to work correctly when exception stack has been installed.

Updated tests to lockdown these scenarios.

File: enclave/core/sgx/td.c [hunk 5]
```diff
@@ -168,6 +168,30 @@ oe_sgx_td_t* oe_sgx_get_td()
...
+*/
+bool oe_sgx_set_td_exception_handler_stack(void* stack, uint64_t size)
+{
+    oe_sgx_td_t* td = oe_sgx_get_td();
+
+    /* ensure the stack + size is 16-byte aligned */
+    if (((uint64_t)stack + size) % 16)
+        return false;
+
+    td->exception_handler_stack_size = size;
```
Comment (anakrish) [inline enclave/core/sgx/td.c:189]: The compiler is free to reorder these assignments and it sometimes it will:
https://godbolt.org/z/EzE7bcd4Y

There needs to be a compiler barrier between the assignments to ensure ordering. Since, assignment is done to thread specific variables that are only used by the current thread, there is no need for an acquire-release pattern.

File: enclave/core/sgx/enter.S (hunk 4 above)
Comment (anakrish) [reply enclave/core/sgx/enter.S:135]: This does not seem to be addressed.

File: tests/VectorException/enc/enc.c [hunk 6]
```diff
@@ -305,21 +353,41 @@ uint64_t test_sigill_handler_with_ocall(oe_exception_record_t* exception_record)
...
+    _use_exception_handler_stack = use_exception_handler_stack;
+
     oe_result_t result = oe_add_vectored_exception_handler(
         false, test_sigill_handler_with_ocall);
     if (result != OE_OK)
     {
         return -1;
     }
 
+    if (_use_exception_handler_stack)
```
Comment (anakrish) [inline tests/VectorException/enc/enc.c:374]: This could be refactored into a separate function.

**Review by anakrish — COMMENTED** [review]
Looks good. Just couple of issues.

File: enclave/core/sgx/enter.S (hunk 4 above)
Comment (mingweishih) [reply enclave/core/sgx/enter.S:135]: Now after all the checks pass, the code directly jumps to `.call_function` without subtracting ABI_REDZONE_BYTES from stack pointer. Is this what's you suggested?

Comment (anakrish) [reply enclave/core/sgx/enter.S:135]: Can you double-check if we need to add an lfence at the fall-through of the exception handler conditional jumps?

File: enclave/core/sgx/td.c [hunk 7]
```diff
@@ -168,6 +168,32 @@ oe_sgx_td_t* oe_sgx_get_td()
...
+    /* Put a barrier to prevent the compiler reorders the operations */
```
Comment (anakrish) [inline enclave/core/sgx/td.c:190]:
```suggestion
    /* Put a barrier to prevent the compiler reordering of the assignment */
```

Comment (anakrish) [reply enclave/core/sgx/td.c:190]: Also document why the order is important. In the comment cross refer to enter.S

File: enclave/core/sgx/enter.S [hunk 8]
```diff
@@ -105,15 +105,42 @@ oe_enter:
...
+    mov td_exception_handler_stack_size(%r11), %r15
+    add %r14, %r15
+    jc .exception_stack_setup // check for overflow
+
+    // Check if the stack range is within the enclave memory range
+    // If the check fails, fallback to the default behavior (i.e.,
+    // re-using the stack pointer saved in the SSA)
+    cmp %r12, %r14
+    jb .exception_stack_setup
+    cmp %r13, %r15
```
Comment (anakrish) [inline enclave/core/sgx/enter.S:131]: Document/describe the scenario where size has been set but stack has not yet been set. That is, when r14 is zero and r15 is 0 + size.

File: enclave/core/sgx/td.c (hunk 5 above)
Comment (mingweishih) [reply enclave/core/sgx/td.c:189]: addressed by checking if the size is zero

File: enclave/core/sgx/td.c (hunk 7 above)
Comment (mingweishih) [reply enclave/core/sgx/td.c]: now we check size != 0, no longer needs this

File: enclave/core/sgx/enter.S (hunk 8 above)
Comment (mingweishih) [reply enclave/core/sgx/enter.S:133]: now we check size != 0, no longer needs this

**Review by anakrish — APPROVED** [review]
LGTM

**Comment by mingweishih** [comment]
bors try

**Comment by mingweishih** [comment]
bors try

**Review by RRathna — APPROVED** [review]

**Review by radhikaj — APPROVED** [review]
Bors r+
